wp-includes/sqlite-ast/class-wp-sqlite-driver.php

```diff
@@ -0,0 +1,2592 @@
+		if ( defined( 'SQLITE_JOURNAL_MODE' ) && in_array( SQLITE_JOURNAL_MODE, $valid_journal_modes, true ) ) {
+			$this->pdo->query( 'PRAGMA journal_mode = ' . SQLITE_JOURNAL_MODE );
+		}
+	}
+
+	/**
+	 * Get the PDO object.
+	 *
+	 * @return PDO
+	 */
+	public function get_pdo() {
+		return $this->pdo;
+	}
+
+	/**
+	 * Method to return inserted row id.
+	 */
+	public function get_insert_id() {
+		return $this->last_insert_id;
+	}
+
+	/**
+	 * Method to return the number of rows affected.
+	 */
+	public function get_affected_rows() {
+		return $this->affected_rows;
+	}
+
+	/**
+	 * Translate and execute a MySQL query in SQLite.
+	 *
+	 * A single MySQL query can be translated into zero or more SQLite queries.
+	 *
+	 * @param string $query              Full SQL statement string.
+	 * @param int    $fetch_mode         PDO fetch mode. Default is PDO::FETCH_OBJ.
+	 * @param array  ...$fetch_mode_args Additional fetch mode arguments.
+	 *
+	 * @return mixed        Return value, depending on the query type.
+	 * @throws Exception    If the query could not run.
+	 * @throws PDOException If the translated query could not run.
+	 */
+	public function query( string $query, $fetch_mode = PDO::FETCH_OBJ, ...$fetch_mode_args ) {
+		$this->flush();
+		if ( function_exists( 'apply_filters' ) ) {
+			/**
+			 * Filters queries before they are translated and run.
+			 *
+			 * Return a non-null value to cause query() to return early with that result.
+			 * Use this filter to intercept queries that don't work correctly in SQLite.
+			 *
+			 * From within the filter you can do
+			 *  function filter_sql ($result, $translator, $statement, $mode, $fetch_mode_args) {
+			 *     if ( intercepting this query  ) {
+			 *       return $translator->execute_sqlite_query( $statement );
+			 *     }
+			 *     return $result;
+			 *   }
+			 *
+			 * @param null|array $result Default null to continue with the query.
+			 * @param object     $translator The translator object. You can call $translator->execute_sqlite_query().
+			 * @param string     $query The statement passed.
+			 * @param int        $fetch_mode Fetch mode: PDO::FETCH_OBJ, PDO::FETCH_CLASS, etc.
+			 * @param array      $fetch_mode_args Variable arguments passed to query.
+			 *
+			 * @returns null|array Null to proceed, or an array containing a resultset.
+			 * @since 2.1.0
+			 */
+			$pre = apply_filters( 'pre_query_sqlite_db', null, $this, $query, $fetch_mode, $fetch_mode_args );
+			if ( null !== $pre ) {
+				return $pre;
+			}
+		}
+		$this->pdo_fetch_mode = $fetch_mode;
+		$this->mysql_query    = $query;
+
+		try {
+			// Parse the MySQL query.
+			$lexer  = new WP_MySQL_Lexer( $query );
+			$tokens = $lexer->remaining_tokens();
+
+			$parser = new WP_MySQL_Parser( self::$grammar, $tokens );
+			$ast    = $parser->parse();
+
+			if ( null === $ast ) {
+				throw new Exception( 'Failed to parse the MySQL query.' );
```

Comment:
Let's subclass the Exception to avoid mixing up our exceptions with potential exceptions from other plugins (e.g. thrown in hooks and filters)